### DISCUSSION TITLE
Padding and packing

### DISCUSSION THREAD

**Discussion by H1manshu21** [Blaze-Stars/Image-Filter-Application#10] [Q&A] [answered]
Must read for bmp.h https://www.geeksforgeeks.org/structure-member-alignment-padding-and-data-packing/ @sabyasachi07 @chandrakant100 @HembramBeta777 @Ank1taS @Tushar7-coder

**Comment by Ank1taS** [comment, marked as answer]
`__attribute__((__packed__))` in [bmp.h](https://github.com/Blaze-Stars/Image-Filter/blob/main/bmp.h) [line 34 and 58] is used to avoid padded bytes among the attributes of structure.

`Padding` aligns structure members with default address boundaries by the processor whereas `packing` prevents the processor from doing padding. Padding makes the size of the struct bigger by applying some rules of memory alignment whereas packing provides the actual size of the struct to which each member is rounded off.
You can refer to [this](https://stackoverflow.com/questions/4306186/structure-padding-and-packing) for more.

- **Rule**: Every member of the struct should be at an address divisible by its size.

##### EXAMPLE 1
```c
struct structure_a {
    long l;
    int i;
    char c;
};
```
|  Member  |  Address  |  Is address multiple of size |
|  ---  |  ---  |  ---  |
|  long  |  0 -7  |  yes (8 * 0)  |
|  int  |  8 - 11  |  yes (4 * 2)  |
|  char  |  12  |  yes (1 * 12)  |

- with padding size of `structure_a = 16`, [8 + 4 + 1, then round to multiple of 8 (long's size)]
- with packing size of `structure_a = 13` 


##### EXAMPLE 2

```c
struct structure_b {
    int i;
    long l;
    char c;
};
```
|  Member  |  Address  |  Is address multiple of size |
|  ---  |  ---  |  ---  |
|  int  |  0-3  |  yes (4 * 0)  |
|  pad  |  4-7  |  |
|  long  |  8-15  |  yes (8 * 1)  |
|  char  |  16  |  yes (1 * 16)  |

- with padding size of `structure_b = 24`, [long need padding by 4 before it, then round to multiple of 8 (long's size)]
- with packing size of `structure_b = 13`

**Reply by H1manshu21** [reply]
Reference for #8. @chandrakant100 @sabyasachi07 @HembramBeta777 @Tushar7-coder

**Reply by chandrakant100** [reply]
done
